pkg/v1beta1/cache/cache.go

```diff
@@ -0,0 +1,166 @@
+package cache
+
+import (
+	"fmt"
+	"strings"
+
+	"sync"
+
+	"github.com/go-logr/logr"
+	v1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	klog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// GatewayLookupCache provides concurrency safe lookups from dns host to namespace/gateway
+// Use New() as the Add, Delete, and Get functions all assume the cache map is non-nil
+type GatewayLookupCache struct {
+	cache  map[string]string
+	mutex  sync.Mutex
+	logger logr.Logger
+}
+
+func New() *GatewayLookupCache {
+	return &GatewayLookupCache{
+		cache:  make(map[string]string),
+		mutex:  sync.Mutex{},
+		logger: klog.Log,
+	}
+
+}
+
+func (glc *GatewayLookupCache) Add(gateway string, hosts ...string) {
+	glc.mutex.Lock()
+	defer glc.mutex.Unlock()
+
+	for _, host := range hosts {
+		glc.cache[host] = gateway
+	}
+}
+
+func (glc *GatewayLookupCache) Delete(hosts ...string) {
+	glc.mutex.Lock()
+	defer glc.mutex.Unlock()
+
+	for _, host := range hosts {
+		delete(glc.cache, host)
+	}
+}
+
+func (glc *GatewayLookupCache) Get(host string) (string, bool) {
+	glc.mutex.Lock()
+	defer glc.mutex.Unlock()
+
+	gw, ok := glc.cache[host]
+	return gw, ok
+}
+
+func (glc *GatewayLookupCache) AddFunc(obj interface{}) {
+	gw, ok := obj.(*v1beta1.Gateway)
+	if !ok {
+		glc.logger.V(1).Info("Not a gateway.v1beta1.istio.io resource")
+		return
+
+	}
+
+	if gw == nil {
+		return
```

Comment:
Should this also Info Log?
```suggestion
                glc.logger.V(1).Info("gateway.v1beta1.istio.io resource nil")
		return
```

Reply:
I figure it is a no-op really. There is nothing to do with a nil pointer here and we have no meaningful metadata to log about it.